File: apo-pot/trunk/src/main/java/com/uqbar/pot/aop/TransactionFieldAccessInterceptor.java
package com.uqbar.pot.aop;

import javassist.Modifier;
import javassist.NotFoundException;
import javassist.expr.FieldAccess;

import org.apache.commons.lang.StringUtils;

import com.uqbar.aop.javassit.parser.Tokens;
import com.uqbar.renascent.framework.aop.FieldAccessInterceptor;


/**
 * @author
 */
public class TransactionFieldAccessInterceptor extends FieldAccessInterceptor {

	@Override
	protected void modifyWriterFieldAccess(StringBuffer statement, FieldAccess fieldAccess) throws NotFoundException {
		if(!Modifier.isTransient(fieldAccess.getField().getModifiers())){
			String reemplaze = Tokens.DEFAULT_VALUE_ASSIGMENT.getRegExp();
			
			String newExpresion = 
				"($fieldTypeName) $interceptor.fieldWrite($this, '$fieldName', $argument1, $this.$fieldName);";
			statement.replace(0, statement.length(), StringUtils.replace(statement.toString(), reemplaze, newExpresion));
		}
	}

	@Override
	protected void modifyReaderFieldAccess(StringBuffer statement, FieldAccess fieldAccess) throws NotFoundException {
		String reemplaze = Tokens.DEFAULT_FIELD.getRegExp();
		
		String newExpresion = "($fieldTypeName) $interceptor.fieldRead($this, '$fieldName', $this.$fieldName);";
		
		statement.replace(0, statement.length(), StringUtils.replace(statement.toString(), reemplaze, newExpresion));
	}

	@Override
	protected String getSpecificPropertyKey() {
		return "TransactionFieldAccessInterceptor";
	}

}

File: apo-pot/trunk/src/main/scala/com/uqbar/pot/aop/TransactionFieldAccessInterceptor.scala
package com.uqbar.pot.aop

import javassist.Modifier
import javassist.NotFoundException
import javassist.expr.FieldAccess
import org.apache.commons.lang.StringUtils
import com.uqbar.aop.javassit.parser.Tokens
import com.uqbar.aop.Interceptor
import com.uqbar.aop.FieldInterceptor

class TransactionFieldInterceptor extends FieldInterceptor {

  write((statement, field) => {
    if (!Modifier.isTransient(field.getField().getModifiers())) {
      var newExpresion =
        """
		  $defaultField = ($fieldTypeName) $interceptor.fieldWrite($this, '$fieldName', $argument1, $this.$fieldName);
		"""
      var reemplaze = Tokens.ORIGINAL_STATEMENT_ASSIGMENT.getRegExp();
      statement.replace(0, statement.length(), StringUtils.replace(statement.toString(), reemplaze, newExpresion));
    }
  })

  read((statement, field) => {
    var reemplaze = Tokens.ORIGINAL_STATEMENT_READER.getRegExp();
    var newExpresion = "$rtn ($fieldTypeName) $interceptor.fieldRead($this, '$fieldName', $this.$fieldName);";
    statement.replace(0, statement.length(), StringUtils.replace(statement.toString(), reemplaze, newExpresion));
  })

  override def getSpecificPropertyKey() = "TransactionFieldAccessInterceptor"
}